storage/innobase/xtrabackup/src/changed_page_tracking.cc

```diff
@@ -0,0 +1,296 @@
+  }
+
+  if (UNIV_UNLIKELY(page_start_lsn > page_end_lsn)) {
+    msg("xtrabackup: pagetracking: incremental backup LSN " LSN_PF
+        " is larger than the last checkpoint LSN " LSN_PF "\n",
+        page_start_lsn, page_end_lsn);
+    return nullptr;
+  }
+
+  unsigned long long int backupid = 0;
+  if (!get_changed_pages(page_start_lsn, page_end_lsn, &backupid, connection)) {
+    msg("xtrabackup: pagetracking: error: Failed to get page tracking file "
+        "from server");
+    return nullptr;
+  }
+
+  File file{-1};
+
+  char full_path[FN_REFLEN];
+
+  fn_format(
+      full_path,
+      (std::string(srv_data_home) + FN_LIBCHAR +
+       Backup_comp_constants::backup_scratch_dir + FN_LIBCHAR +
+       std::to_string(backupid) + Backup_comp_constants::change_file_extension)
+          .c_str(),
+      "", "", MY_RETURN_REAL_PATH);
+
+  Fil_path::normalize(full_path);
+
+  file = my_open(full_path, O_RDONLY, MYF(MY_WME));
+
+  if (file < 0) {
+    msg("xtrabackup: pagetracking: error: cannot open '%s'\n", full_path);
+    return nullptr;
+  }
+
+  //  auto file_size = os_file_get_size(filename.c_str());
+  /* todo: Introduce new option to limit the size of file we can process?  */
+
+  aligned_array_pointer<byte, UNIV_PAGE_SIZE_MAX> buf;
+  buf.create(page_tracking_read_buffer_size);
+  size_t hdr_len = page_tracking_read_buffer_size;
+
+  space_map = new xb_space_map;
+
+  /* process file to create map of space and it's changed pages */
+  while (true) {
+    size_t n_read = my_read(file, buf, hdr_len, MYF(MY_WME));
+
+    ut_a(n_read % Backup_comp_constants::page_number_size == 0);
+
+    if (n_read == MY_FILE_ERROR) {
+      msg("xtrabackup: cannot read from '%s'\n", full_path);
+      my_close(file, MYF(MY_FAE));
+      return nullptr;
+    }
+
+    space_id_t space_id;
+    page_no_t page_id;
+
+    auto num_pages = n_read / Backup_comp_constants::page_number_size;
+
+    for (size_t index = 0; index < num_pages; index++) {
+      space_id = mach_read_from_4(
+          buf + Backup_comp_constants::page_number_size * index);
+      page_id =
+          mach_read_from_4(buf + Backup_comp_constants::page_number_size / 2 +
+                           Backup_comp_constants::page_number_size * index);
+
+      /* insert page_id into existing space or create new space and insert */
+      if (space_map->count(space_id)) {
+        space_map->at(space_id).insert(page_id);
+      } else {
+        space_map->insert({space_id, {page_id}});
+      }
+    }
+
+    if (n_read < hdr_len) {
+      break;
+    }
+  }
+
+  my_close(file, MYF(MY_FAE));
+  return space_map;
+}
+
+/** Free the tracking map.
+@param[in/out] tracking map */
+void deinit(xb_space_map *space_map) {
+  if (space_map) {
+    delete space_map;
+  }
+}
+
+/** Call the mysqlbackup component mysqlbackup_page_track_get_changed_pages
+to get the pages between the LSN interval incremental_lsn to
+checkpoint_lsn_start. Server writes file havinng spaceid and page in data
```

Comment:
Server writes file containing*

Reply:
Or  Server writes files with space_id and page_num

Reply:
fixed